app/src/main/java/com/duckduckgo/app/browser/favicon/FaviconManager.kt

```diff
@@ -0,0 +1,186 @@
+    suspend fun loadToViewFromTemp(subFolder: String, url: String, view: ImageView)
+    suspend fun loadToViewFromPersisted(url: String, view: ImageView)
+    suspend fun deletePersistedFavicon(url: String)
+    suspend fun deleteOldTempFavicon(subFolder: String, path: String?)
+    suspend fun deleteAllTemp()
+}
+
+class DuckDuckGoFaviconManager constructor(
+    private val faviconPersister: FaviconPersister,
+    private val bookmarksDao: BookmarksDao,
+    private val fireproofWebsiteRepository: FireproofWebsiteRepository,
+    private val locationPermissionsRepository: LocationPermissionsRepository,
+    private val faviconDownloader: FaviconDownloader,
+    private val dispatcherProvider: DispatcherProvider
+) : FaviconManager {
+
+    override suspend fun loadFromTemp(subFolder: String, url: String): Bitmap? {
+        val domain = extractDomain(url)
+        val cachedFavicon = faviconPersister.faviconFile(FAVICON_TEMP_DIR, subFolder, domain)
+        return if (cachedFavicon != null) {
+            faviconDownloader.getFaviconFromDisk(cachedFavicon)
+        } else {
+            null
+        }
+    }
+
+    override suspend fun loadToViewFromPersisted(url: String, view: ImageView) {
+        loadToViewFromDirectory(url, FAVICON_PERSISTED_DIR, NO_SUBFOLDER, view)
+    }
+
+    override suspend fun loadToViewFromTemp(subFolder: String, url: String, view: ImageView) {
+        loadToViewFromDirectory(url, FAVICON_TEMP_DIR, subFolder, view)
+    }
+
+    override suspend fun prefetchToTemp(subFolder: String, url: String): File? {
+        val domain = url.toUri().domain() ?: return null
+        val favicon = downloadFromUrl(url)
+        return if (favicon != null) {
+            saveToFile(FAVICON_TEMP_DIR, subFolder, favicon, domain)
+        } else {
+            null
+        }
+    }
+
+    override suspend fun saveToTemp(subFolder: String, icon: Bitmap, url: String): File? {
+        val domain = extractDomain(url)
+        val file = faviconPersister.store(FAVICON_TEMP_DIR, subFolder, icon, domain)
+        if (file != null) { // Only replace persisted favicons if we stored a new file
+            replacePersistedFavicons(icon, domain)
+        }
+        return file
+    }
+
+    override suspend fun persistFavicon(subFolder: String, url: String) {
+        val domain = extractDomain(url)
+        val cachedFavicon = faviconPersister.faviconFile(FAVICON_TEMP_DIR, subFolder, domain)
+        if (cachedFavicon != null) {
+            faviconPersister.copyToDirectory(cachedFavicon, FAVICON_PERSISTED_DIR, NO_SUBFOLDER, domain)
+        }
+    }
+
+    override suspend fun deletePersistedFavicon(url: String) {
+        val domain = extractDomain(url)
+        val remainingFavicons = persistedFaviconsForDomain(domain) - 1 // -1 because we are about to delete one
+        if (remainingFavicons <= 0) {
+            faviconPersister.deletePersistedFavicon(domain)
+        }
```

Comment:
I think this would be clearer if you didn't subtract that minus 1 from the first line; i think it's a little confusing. is the logic essentially to check if there is only a single instance of the favicon, and delete it if so? so the logic boils down to `numberFavicons == 1`, or am i missing something? do you need to call `delete` if you find 0 favicons are returned from that count?

```
val remainingFavicons = persistedFaviconsForDomain(domain)
if (remainingFavicons == 1) {
    faviconPersister.deletePersistedFavicon(domain)
}
```

Reply:
Makes sense to me, changed to `== 1`